azure-keyvault-cryptography/src/main/java/com/microsoft/azure/keyvault/cryptography/algorithms/PsBase.java

```diff
@@ -0,0 +1,97 @@
+import java.security.NoSuchAlgorithmException;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.PSSParameterSpec;
+import java.util.Arrays;
+import java.util.zip.DataFormatException;
+import java.security.Signature;
+
+import com.microsoft.azure.keyvault.cryptography.ByteExtensions;
+import com.microsoft.azure.keyvault.cryptography.ISignatureTransform;
+
+/**
+ *
+ */
+public abstract class PsBase extends RsaSignature {
+
+    protected PsBase(String name) {
+        super(name);
+    }
+
+    class PsBaseSignatureTransform implements ISignatureTransform {
+
+        private final KeyPair _keyPair;
+        private final int _emLen;
+        private final int _modBits;
+        private final String _digestName;
+
+        PsBaseSignatureTransform(KeyPair keyPair, String digestName) {
+            _keyPair = keyPair;
+            BigInteger modulus = ((RSAPublicKey) _keyPair.getPublic()).getModulus();
+            _emLen = getOctetLength(modulus.bitLength());
+            _modBits = ((RSAPublicKey) _keyPair.getPublic()).getModulus().bitLength();
+            _digestName = digestName;
+        }
+
+        @Override
+        public byte[] sign(byte[] digest) throws NoSuchAlgorithmException {
+
+            if (_keyPair.getPrivate() == null) {
+                throw new IllegalArgumentException("Keypair must have private key to for signing.");
+            }
+
+            // Construct the encoded message
+            // Apply the EMSA-PSS encoding operation (Section
+            // 9.1.1) to the message M to produce an encoded message EM of length
+            // \ceil ((modBits - 1)/8) octets such that the bit length of the
+            // integer OS2IP (EM) (see Section 4.2) is at most modBits - 1, where
+            // modBits is the length in bits of the RSA modulus n:
+            MessageDigest messageDigest = MessageDigest.getInstance(_digestName);
+            byte[] EM = EMSA_PSS_ENCODE_HASH(digest, _modBits - 1, messageDigest);
+
+            // Convert to integer message
+            BigInteger m = OS2IP(EM);
+
+            // RSASP1(s)
+            m = RSASP1((RSAPrivateKey) _keyPair.getPrivate(), m);
+
+            // Convert to octet sequence
+            return I2OSP(m, _emLen);
+        }
+
+        @Override
+        public boolean verify(byte[] digest, byte[] signature) throws NoSuchAlgorithmException {
+
+            if (signature.length != _emLen) {
+                throw new IllegalArgumentException("invalid signature length");
+            }
+
+            // Convert to integer signature
+            BigInteger s = OS2IP(signature);
+
+            // Convert integer message
+            BigInteger m = RSAVP1((RSAPublicKey) _keyPair.getPublic(), s);
+
+            byte[] EM = I2OSP(m, _emLen);
+
+            MessageDigest messageDigest = MessageDigest.getInstance(_digestName);
+            return EMSA_PSS_VERIFY(digest, EM, _modBits, messageDigest);
```

Comment:
>_modBits [](start = 47, length = 8)

"Result = EMSA-PSS-VERIFY (M, EM, modBits - 1)". missing the - 1

Reply:
EMSA-PSS-VERIFY is actually supposed to use emBits, not modBits. emBits is the bit length of EM, and in the ctor of we call getOctetLength to get _emLen which is the octet length of EM. In theory, emBits should always be the same as _emLen / 8.